File: test/openalea/shootnumber.py
import numpy as np 
from copy import copy
from math import *

from fibonacci import * 
def shootnumber(canopyShootNumber=288.0,
                leafNumber=0.0,
                sowingDensity=288,
                targetFertileShoot=600.0,
                tilleringProfile=[288],
                leafTillerNumberArray=[1],
                tillerNumber=1):
    """


    CalculateShootNumber Model
    Author: Pierre MARTRE
    Reference: Modeling development phase in the 
                Wheat Simulation Model SiriusQuality.
                See documentation at http://www1.clermont.inra.fr/siriusquality/?page_id=427
    Institution: INRA/LEPSE Montpellier
    Abstract: calculate the shoot number and update the related variables if needed

    """
    oldCanopyShootNumber = canopyShootNumber
    emergedLeaves = int(max(1, ceil(leafNumber - 1)))
    shoots = fibonacci(emergedLeaves)
    canopyShootNumber = min(shoots * sowingDensity, targetFertileShoot)
    averageShootNumberPerPlant = canopyShootNumber / sowingDensity       
    if (canopyShootNumber != oldCanopyShootNumber):
        tilleringProfile.append(canopyShootNumber - oldCanopyShootNumber)         
    tillerNumber = len(tilleringProfile)     
    for i in range(len(leafTillerNumberArray),int(ceil(leafNumber))):
        leafTillerNumberArray.append(tillerNumber)
    return  averageShootNumberPerPlant, canopyShootNumber, leafTillerNumberArray, tilleringProfile, tillerNumber

File: src/cpp/source/ShootNumber.cpp
#include <iostream>
#include <vector>
#include<string>
#include<algorithm>
#include<cmath>
#include<time.h>
#include "Fibonacci.h"
#include "ShootNumber.h"

using namespace std;
ShootNumber Calculate_ShootNumber(double canopyShootNumber,double leafNumber,int sowingDensity,double targetFertileShoot,vector<double> tilleringProfile,vector<int> leafTillerNumberArray,int tillerNumber)
{
    double averageShootNumberPerPlant;
    double oldCanopyShootNumber;
    int emergedLeaves,shoots;
    oldCanopyShootNumber= canopyShootNumber;
    emergedLeaves = (int)fmax(1, ceil(leafNumber - 1));
    shoots = Calculate_Fibonacci(emergedLeaves).value;
    canopyShootNumber = fmin(shoots * sowingDensity, targetFertileShoot);
    averageShootNumberPerPlant = canopyShootNumber / sowingDensity;
    if (canopyShootNumber != oldCanopyShootNumber)
    {
        tilleringProfile.push_back(canopyShootNumber - oldCanopyShootNumber);
    }
    tillerNumber = tilleringProfile.size();
    for (int i = leafTillerNumberArray.size(); i < ceil(leafNumber); i++)
    {
        leafTillerNumberArray.push_back(tillerNumber);
    }
    return ShootNumber{averageShootNumberPerPlant, canopyShootNumber,leafTillerNumberArray,tilleringProfile,tillerNumber};
}
